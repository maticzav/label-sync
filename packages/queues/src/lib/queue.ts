import crypto from 'crypto'
import pino from 'pino'
import * as redis from 'redis'

import { sleep, UnionOmit } from './utils'

/**
 * The base type to use to build a new task.
 */
export type TaskSpec = {
  id: string

  /**
   * IDs of other tasks this task depends on.
   *
   * NOTE: If any of the tasks in this array is still in the queue,
   *       the system is going to put this task to the back of the queue again.
   */
  dependsOn: string[]
}

/**
 * A general purpose queue backed by Redis.
 */
export class Queue<Task extends TaskSpec> {
  /**
   * Name of the queue.
   */
  private name: string

  /**
   * Client to use for communication with the Redis queue.
   */
  private client: redis.RedisClientType

  /**
   * An intenral logger that the queue may use to log events.
   */
  private logger: pino.Logger

  /**
   * Number of functions that are currently processing tasks.
   */
  private locks: number = 0

  /**
   * Tells whether the queue has stopped processing requests.
   */
  private disposed: boolean = false

  /**
   * Functions that should be called when the queue is successfully disposed.
   */
  private releases: (() => void)[] = []

  private connected: boolean = false

  constructor(name: string, url: string) {
    this.name = name
    this.logger = pino()

    this.logger.info(`Connecting to Redis at "${url}"!`)

    this.client = redis.createClient({ url })
    this.client.on('error', (err) => this.logger.error(err, 'Redis Client Error'))
  }

  /**
   * Starts the queue and connects to Redis.
   */
  public async start(): Promise<void> {
    this.logger.info(`Connecting to queue "${this.name}"...`)
    if (!this.connected) {
      await this.client.connect()
    }
    this.connected = true
    this.logger.info(`Connected to queue "${this.name}"!`)
  }

  /**
   * Returns the list identifier for the queue.
   */
  private queue(): string {
    return `queue:${this.name}`
  }

  /**
   * Returns the name of the set used to list unprocessed tasks.
   */
  private members(): string {
    return `queuemembers:${this.name}`
  }

  /**
   * Lists all tasks that are currently in the queue.
   */
  public async list(): Promise<Task[]> {
    if (!this.connected) {
      await this.start()
    }

    this.logger.info(`Listing tasks in queue "${this.name}"!`)

    const rawtasks = await this.client.LRANGE(this.queue(), 0, -1)
    const tasks = rawtasks.map((raw) => JSON.parse(raw) as Task)

    return tasks
  }

  /**
   * Function that pushes a new task to the queue and returns
   * the task identifier.
   */
  public async push(task: UnionOmit<Task, 'id'>): Promise<string> {
    if (!this.connected) {
      await this.start()
    }

    const id = crypto.randomUUID()
    const hydratedTask = { id, ...task } as Task

    return this._push(hydratedTask)
  }

  /**
   * Adds a complete task to the queue.
   */
  private async _push(task: Task): Promise<string> {
    this.logger.info(task, `Pushing task "${task.id}" to queue "${this.name}"!`)

    await Promise.all([
      this.client.RPUSH(this.queue(), JSON.stringify(task)),
      this.client.SADD(this.members(), task.id),
    ])

    return task.id
  }

  /**
   * A function that waits for new task in the queue and forwards it
   * to the executor for processing.
   */
  public async process(fn: (task: Task) => Promise<boolean>) {
    if (!this.connected) {
      await this.start()
    }

    processor: while (!this.disposed) {
      this.logger.info(`Checking for new task in queue "${this.name}"!`)

      const rawtask = await this.client.BLPOP([this.queue()], 30)
      if (!rawtask) {
        if (!this.disposed) {
          await sleep(100)
          continue
        }

        // If we're disposed, we're done and stop processing tasks.
        return
      }
      const task = JSON.parse(rawtask.element) as Task

      // If some of the task's dependencies haven't been processed yet,
      // process them first. The for loop is synchronous because the
      // task usually only depends on a few dependencies and it
      // doesn't add significant overhead to the processing.
      for (const dependencyId of task.dependsOn) {
        const unprocessed = await this.client.SISMEMBER(this.members(), dependencyId)
        if (unprocessed) {
          await this._push(task)
          continue processor
        }
      }

      // We process the task and add it back to the queue in case
      // something went wrong.
      try {
        this.locks++
        const more = await fn(task)
        await this.client.SREM(this.members(), task.id)
        this.locks--

        await this._dispose()

        if (!more || this.disposed) {
          return
        }
      } catch {
        await this._push(task)
      }
    }
  }

  /**
   * Intenal utility function that disposes the client.
   */
  private async _dispose() {
    if (this.locks > 0 || !this.disposed) {
      return
    }

    await this.client.disconnect()
    this.locks = -1
    for (const release of this.releases) {
      release()
    }
  }

  /**
   * Disposes the client and returns a promise that resolves when
   * the connection is dropped.
   */
  public async dispose(): Promise<void> {
    this.disposed = true

    if (this.locks > 0) {
      return new Promise<void>((resolve) => {
        this.releases.push(resolve)
      })
    }

    await this._dispose()

    return
  }
}
