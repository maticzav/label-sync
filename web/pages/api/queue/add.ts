import { withAuth, WithAuthProp, users } from '@clerk/nextjs/api'
import type { NextApiRequest, NextApiResponse } from 'next'
import * as z from 'zod'

import * as tasks from 'lib/tasks'

const schema = z.discriminatedUnion('kind', [
  z.object({
    kind: z.literal('onboard_org'),
    ghInstallationId: z.number(),
    org: z.string(),
    accountType: z.string(),
  }),
  z.object({
    kind: z.literal('sync_org'),
    ghInstallationId: z.number(),
    isPaidPlan: z.boolean(),

    org: z.string(),
  }),
  z.object({
    kind: z.literal('sync_repo'),
    ghInstallationId: z.number(),
    isPaidPlan: z.boolean(),

    repo: z.string(),
    org: z.string(),
  }),
  z.object({
    kind: z.literal('dryrun_config'),
    ghInstallationId: z.number(),
    isPaidPlan: z.boolean(),

    org: z.string(),
    pr_number: z.number(),
  }),
  z.object({
    kind: z.literal('add_siblings'),
    ghInstallationId: z.number(),
    isPaidPlan: z.boolean(),

    repo: z.string(),
    org: z.string(),

    issue_number: z.number(),
    label: z.string(),
  }),
  z.object({
    kind: z.literal('check_unconfigured_labels'),
    ghInstallationId: z.number(),
    isPaidPlan: z.boolean(),

    repo: z.string(),
    org: z.string(),

    label: z.string(),
  }),
])

export default withAuth(async (req: WithAuthProp<NextApiRequest>, res: NextApiResponse) => {
  if (!req.auth.userId) {
    res.status(403).json({ message: 'Unauthorized' })
    return
  }

  const user = await users.getUser(req.auth.userId)
  if (!user.publicMetadata['is_admin']) {
    res.status(403).json({ message: 'Unauthorized' })
    return
  }

  const parsed = schema.safeParse(JSON.parse(req.body))

  if (!parsed.success) {
    res.status(400).json({ message: parsed.error.message })
    return
  }

  const id = await tasks.shared.push({
    ...parsed.data,
    dependsOn: [],
  })

  res.status(200).json({ id })
})
