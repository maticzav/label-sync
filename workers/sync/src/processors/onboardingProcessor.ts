import { getLSConfigRepoName } from '@labelsync/config'

import { Processor } from '../lib/processor'
import { populateTemplate, TEMPLATES } from '../lib/templates'

type ProcessorData = {
  owner: string
  /**
   * The type of the account ("organization", "personal").
   */
  accountType: string
}

/**
 * Processor that onboards an organization to the platform.
 */
export class OnboardingProcessor extends Processor<ProcessorData> {
  async perform({ owner, accountType }: ProcessorData): Promise<void> {
    this.log.info(`Onboarding "${owner}"!`)

    const configRepoName = getLSConfigRepoName(owner)

    const accessibleRepos = await this.endpoints.checkInstallationAccess({
      owner,
      repos: [],
    })
    if (accessibleRepos == null) {
      throw new Error(`Couldn't check repository access.`)
    }

    // If configuration repository already exists, there's no need to bootstrap anything.
    const repo = await this.endpoints.getRepo({ owner, repo: configRepoName })
    if (repo != null) {
      this.log.info(`Configuration repository already exists, skipping onboarding.`)
      return
    }

    this.log.info(`No existing repository for ${owner}, start onboarding!`)

    switch (accountType) {
      case 'Organization': {
        this.log.info(`Bootstraping config repo for ${owner}.`)

        // Bootstrap a configuration repository in organisation.
        const repositories = accessibleRepos.accessible.map((name) => ({ name }))
        const personalisedTemplate = populateTemplate(TEMPLATES.yaml, {
          repository: configRepoName,
          repositories,
        })
        await this.endpoints.bootstrapConfigRepository({ owner, repo: configRepoName }, personalisedTemplate)

        this.log.info(`Onboarding complete for ${owner}.`)
        return
      }

      /* istanbul ignore next */
      case 'User': {
        // TODO: Allow personal account scaffolding once Github provides support.
        this.log.info(`User account ${owner}, skip onboarding.`)
        return
      }

      /* istanbul ignore next */
      default: {
        this.log.warn(`Unsupported account type: ${accountType}`)
        return
      }
    }
  }
}
