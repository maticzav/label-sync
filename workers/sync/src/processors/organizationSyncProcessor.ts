import { getLSConfigRepoName, getPhysicalRepositories, parseConfig } from '@labelsync/config'

import { messages } from '../lib/constants'
import { Processor } from '../lib/processor'

type ProcessorData = {
  owner: string
}

/**
 * A processor that verifies configuration and syncs the configuration with
 * the labels configured in each configured repository.
 */
export class OrganizationSyncProcessor extends Processor<ProcessorData> {
  /**
   * Syncs configuration of a repository with its labels.
   */
  public async perform({ owner }: ProcessorData) {
    const configRepoName = getLSConfigRepoName(owner)
    const rawConfig = await this.endpoints.getConfig({ owner })

    /* No configuration, skip the evaluation. */
    /* istanbul ignore next */
    if (rawConfig === null) {
      this.log.info(`No configuration, skipping sync.`)
      return
    }
    const parsedConfig = parseConfig({ input: rawConfig, isPro: this.installation.isPaidPlan })

    if (!parsedConfig.ok) {
      this.log.info(`Error in configuration, openning issue.`, {
        meta: { config: rawConfig, error: parsedConfig.error },
      })

      const title = 'Configuration Issue'
      const body = messages['onboarding.error.issue'](parsedConfig.error)
      const issue = await this.endpoints.openIssue({ owner, repo: configRepoName }, { title, body })
      if (issue == null) {
        throw new Error(`Couldn't open an issue...`)
      }

      this.log.info(`Opened issue ${issue.id} in ${owner}/${configRepoName}.`)

      return
    }

    // Verify that we can access all configured files.
    const requiredRepositories = getPhysicalRepositories(parsedConfig.config)
    const access = await this.endpoints.checkInstallationAccess({
      owner,
      repos: requiredRepositories,
    })

    if (access == null) {
      throw new Error(`Couldn't check repository access.`)
    }

    if (access.status === 'Sufficient') {
      // Even if there's no wildcard configuration, we try to sync all accessible repositories.
      for (const repo of access.accessible) {
        this.queue.push({
          kind: 'sync_repo',
          repo: repo,
          org: owner,
          dependsOn: [],
          ghInstallationId: this.installation.id,
          isPaidPlan: this.installation.isPaidPlan,
        })
      }
      return
    }

    this.log.info(`Insufficient permissions, skipping sync.`, {
      meta: { access: JSON.stringify(access) },
    })

    const title = 'Insufficient Permissions'
    const body = messages['insufficient.permissions.issue'](access.missing)

    const issue = await this.endpoints.openIssue({ owner, repo: configRepoName }, { title, body })
    if (issue == null) {
      throw new Error(`Couldn't open an issue...`)
    }

    this.log.info(`Opened issue ${issue.id} in ${owner}/${configRepoName}.`)

    return
  }
}
