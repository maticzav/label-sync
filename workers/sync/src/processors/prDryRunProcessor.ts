import { getLSConfigRepoName, getPhysicalRepositories, parseConfig } from '@labelsync/config'
import { calculateConfigurationDiff } from '../lib/config'

import { messages } from '../lib/constants'
import { Processor } from '../lib/processor'
import { generateHumanReadablePRReport, LabelSyncReport } from '../lib/reports'

type ProcessorData = {
  owner: string
  pr_number: number
  isPro: boolean
}

/**
 * A processor that performs a dryrun of the new configuration and comments
 * results on the PR.
 */
export class DryRunProcessor extends Processor<ProcessorData> {
  public async perform({ owner, pr_number, isPro }: ProcessorData) {
    this.log.info(`Performing dryrun for ${owner}/${pr_number}`)

    const repo = getLSConfigRepoName(owner)

    const check = await this.endpoints.createPRCheckRun(
      { repo: repo, owner },
      {
        name: 'LabelSync Configuration Check',
        pr_number,
      },
    )

    const result = await this.dryrun({ owner, repo, isPro, pr_number: pr_number })

    // Don't complete the check if we couldn't create it.
    if (!check) {
      return
    }

    // Complete the check with status if it was created successfully.
    await this.endpoints.completePRCheckRun(
      { owner, repo },
      {
        check_run: check.id,
        conclusion: result.ok ? 'success' : 'failure',
        output: {
          title: 'LabelSync Configuration Check',
          summary: result.message,
          text: result.message,
        },
      },
    )
  }

  /**
   * Performs configuration validation and returns result.
   */
  private async dryrun({
    owner,
    repo,
    isPro,
    pr_number: pr_number,
  }: {
    repo: string
    owner: string
    isPro: boolean
    pr_number: number
  }) {
    const pr = await this.endpoints.getPullRequest({ owner, repo }, { pr_number })
    if (pr == null) {
      this.log.error({ owner, repo, pr_number }, `Could not find pull request for ${owner}/${repo}#${pr_number}`)
      throw new Error(`Could not find PR ${pr_number}`)
    }

    const rawConfig = await this.endpoints.getConfig({ owner, ref: pr.head.ref })
    if (rawConfig === null) {
      this.log.info({ owner, pr }, `No configuration, skipping dryrun.`)
      return { ok: false, message: 'No configuration found.' }
    }

    const parsedConfig = parseConfig({ input: rawConfig, isPro })

    if (!parsedConfig.ok) {
      this.log.info(
        { meta: { config: rawConfig, error: parsedConfig.error } },
        `Error in configuration, openning issue.`,
      )

      const comment = messages['invalid.config.comment'](parsedConfig.error)
      await this.endpoints.comment({ owner, repo }, { issue: pr_number, comment })

      this.log.info({ comment }, `Commented on the issue ${pr_number} in ${owner}/${repo}.`)

      return { ok: false, message: parsedConfig.error }
    }

    // Verify that we can access all configured files.
    const repositories = getPhysicalRepositories(parsedConfig.config)
    const access = await this.endpoints.checkInstallationAccess({ owner, repos: repositories })
    if (access == null) {
      throw new Error(`Couldn't check repository access.`)
    }

    if (access.status === 'Insufficient') {
      this.log.info(`Insufficient permissions, skipping sync.`, {
        meta: { access: JSON.stringify(access) },
      })

      const comment = messages['insufficient.permissions.comment'](access.missing)
      await this.endpoints.comment({ owner, repo }, { issue: pr_number, comment })
      this.log.info({ comment }, `Commented on the issue ${pr_number} in ${owner}/${repo}.`)

      return { ok: false, message: comment }
    }

    // Perform the dryrun on each repository.
    const reports: LabelSyncReport[] = []
    for (const repo of access.accessible) {
      const config = parsedConfig.config.repos[repo.toLowerCase()] ?? parsedConfig.config.repos['*']
      if (config == null) {
        continue
      }

      const labels = await this.endpoints.getLabels({ owner, repo })
      if (labels == null) {
        throw new Error(`Couldn't fetch labels for ${owner}/${repo}.`)
      }

      const dryrun = calculateConfigurationDiff({
        config: config.labels,
        currentLabels: labels,
      })

      reports.push({
        status: 'Success',
        additions: dryrun.added,
        updates: dryrun.changed,
        aliases: dryrun.aliased,
        removals: dryrun.removed,
        repo,
        owner,
        config: {
          labels: config.labels,
          config: {
            removeUnconfiguredLabels: config.config?.removeUnconfiguredLabels === true,
          },
        },
      })
    }

    const report = generateHumanReadablePRReport(reports)
    return { ok: true, message: report }
  }
}
