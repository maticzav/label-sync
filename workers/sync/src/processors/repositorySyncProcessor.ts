import { parseConfig } from '@labelsync/config'

import { calculateConfigurationDiff } from '../lib/config'
import { Processor } from '../lib/processor'
import { famap } from '../lib/utils'

type ProcessorData = {
  owner: string
  repo: string
}

/**
 * A processor that syncs the configuration with the labels configured in a repository.
 */
export class RepositorySyncProcessor extends Processor<ProcessorData> {
  /**
   * Syncs configuration of a repository with its labels.
   */
  public async perform({ owner, repo }: ProcessorData): Promise<void> {
    this.log.info(`Syncing repository ${owner}/${repo}...`)

    const rawConfig = await this.endpoints.getConfig({ owner })
    if (rawConfig === null) {
      this.log.info(`No configuration, skipping siblings sync.`)
      return
    }
    const parsedConfig = parseConfig({
      input: rawConfig,
      isPro: this.installation.isPaidPlan,
    })
    if (!parsedConfig.ok) {
      this.log.info(`Invalid configuration file in organization ${owner}...`)
      return
    }

    const config = parsedConfig.config.repos
    const repoconfig = config[repo.toLowerCase()] ?? config['*']
    if (repoconfig == null) {
      this.log.info(`No configuration for repository ${owner}/${repo}, skipping sync.`)
      return
    }

    const removeUnconfiguredLabels = repoconfig.config?.removeUnconfiguredLabels === true

    const remoteLabels = await this.endpoints.getLabels({ repo, owner })
    if (remoteLabels == null) {
      throw new Error(`Couldn't fetch current labels for repository ${owner}/${repo}.`)
    }

    const { added, changed, aliased, removed } = calculateConfigurationDiff({
      config: repoconfig.labels,
      currentLabels: remoteLabels,
    })

    try {
      await famap(added, (label) => this.endpoints.createLabel({ owner, repo }, label))
      await famap(changed, (label) => this.endpoints.updateLabel({ owner, repo }, label))
      await this.endpoints.aliasLabels({ owner, repo }, aliased)

      if (removeUnconfiguredLabels) {
        await famap(removed, (label) => this.endpoints.removeLabel({ owner, repo }, label))
      }

      this.log.info(`Sync of ${owner}/${repo} completed.`)
    } catch (err: any) /* istanbul ignore next */ {
      this.log.error(err, `Something went wrong during the sync of ${owner}/${repo}`)
    }

    return
  }
}
