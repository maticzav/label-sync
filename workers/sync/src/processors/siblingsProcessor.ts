import { parseConfig } from '@labelsync/config'
import _ from 'lodash'

import { Processor } from '../lib/processor'

type ProcessorData = {
  owner: string
  repo: string
  issue_number: number
  label: string
  isPro: boolean
}

export class SiblingsProcessor extends Processor<ProcessorData> {
  /**
   * Syncs the configuration of a repository with its labels.
   */
  public async perform({ owner, repo, label, issue_number, isPro }: ProcessorData): Promise<void> {
    this.log.info(`${issue_number} labeled with "${label}" in ${owner}/${repo}...`)

    const rawConfig = await this.endpoints.getConfig({ owner })
    if (rawConfig === null) {
      this.log.info(`No configuration, skipping siblings sync.`)
      return
    }
    const parsedConfig = parseConfig({ input: rawConfig, isPro })

    if (!parsedConfig.ok) {
      this.log.info(parsedConfig, `Invalid configuration, skipping siblings sync.`)
      return
    }

    this.log.info(parsedConfig, `Successfully parsed organization configuration!`)
    const config = parsedConfig.config.repos[repo]

    /* istanbul ignore if */
    if (!config) {
      this.log.info(`No configuration found for the repo, skipping.`)
      return
    }

    if (label in config.labels) {
      const siblings = _.get(config, ['labels', label, 'siblings'], [] as string[])

      /* istanbul ignore if */
      if (siblings.length === 0) {
        this.log.info(`No siblings to add to "${label}", skipping.`)
        return
      }

      const labels = siblings.map((sibling) => ({ name: sibling }))
      await this.endpoints.addLabelsToIssue({ repo, owner }, { issue_number, labels })

      return
    }

    this.log.info(`Unconfigured label "${label}", nothing to do.`)
    return
  }
}
