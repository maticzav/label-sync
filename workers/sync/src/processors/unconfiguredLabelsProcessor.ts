import { parseConfig } from '@labelsync/config'
import _ from 'lodash'

import { Processor } from '../lib/processor'

type ProcessorData = {
  owner: string
  repo: string
  label: string
}

/**
 * A processor that removes a label that is not configured in strict repositories.
 */
export class UnconfiguredLabelsProcessor extends Processor<ProcessorData> {
  /**
   * Syncs the configuration of a repository with its labels.
   */
  public async perform({ owner, repo, label }: ProcessorData) {
    this.log.info(`New label created in ${repo}: "${label}".`)

    const rawConfig = await this.endpoints.getConfig({ owner })
    if (rawConfig === null) {
      this.log.info(`No configuration, skipping siblings sync.`)
      return
    }
    const parsedConfig = parseConfig({
      input: rawConfig,
      isPro: this.installation.isPaidPlan,
    })

    if (!parsedConfig.ok) {
      this.log.info(parsedConfig, `Invalid configuration, skipping siblings sync.`)
      return
    }

    const config = parsedConfig.config.repos[repo]

    /* istanbul ignore if */
    if (!config) {
      this.log.info(`No configuration found for the repo, skipping.`)
      return
    }

    if (label in config.labels) {
      this.log.info(`Label "${label}" is configured, skipping.`)
      return
    }

    if (config.config?.removeUnconfiguredLabels === true) {
      this.log.info(`Removing "${label}" from ${repo} because it's not configured.`)
      await this.endpoints.removeLabel({ repo, owner }, { name: label })
      this.log.info(`Removed label "${label}" from ${repo}.`)
    }

    return
  }
}
