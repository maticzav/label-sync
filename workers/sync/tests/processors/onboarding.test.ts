import pino from 'pino'
import { OnboardingProcessor } from '../../src/processors/onboardingProcessor'
import { MockGitHubEndpoints } from '../__fixtures__/endpoints'
import { MockTaskQueue } from '../__fixtures__/queue'

describe('onboarding', () => {
  test('onboards organization', async () => {
    const installation = { id: 1, isPaidPlan: false }
    const endpoints = new MockGitHubEndpoints({
      configs: {},
      installations: {
        'test-org': ['a', 'b', 'c'],
      },
      repos: {},
    })
    const queue = new MockTaskQueue()
    const logger = pino()

    const processor = new OnboardingProcessor(installation, queue, endpoints, logger)

    await processor.perform({
      accountType: 'Organization',
      owner: 'test-org',
    })

    expect(endpoints.stack).toMatchSnapshot()
  })

  test('skips onboarding when config repo exists', async () => {
    const installation = {
      id: 1,
      isPaidPlan: false,
    }
    const endpoints = new MockGitHubEndpoints({
      configs: {},
      installations: {
        'test-org': ['a', 'b', 'c'],
      },
      repos: {
        'test-org': {
          'test-org-labelsync': {
            id: 12,
            default_branch: 'master',
          },
        },
      },
    })
    const queue = new MockTaskQueue()
    const logger = pino()

    const processor = new OnboardingProcessor(installation, queue, endpoints, logger)

    await processor.perform({
      accountType: 'Organization',
      owner: 'test-org',
    })

    expect(endpoints.stack).toMatchSnapshot()
  })

  test('onboards personal account', async () => {
    const installation = {
      id: 1,
      isPaidPlan: false,
    }
    const endpoints = new MockGitHubEndpoints({
      configs: {},
      installations: {
        'test-user': ['a', 'b', 'c'],
      },
    })
    const queue = new MockTaskQueue()
    const logger = pino()

    const processor = new OnboardingProcessor(installation, queue, endpoints, logger)

    await processor.perform({
      accountType: 'User',
      owner: 'test-user',
    })

    expect(endpoints.stack).toMatchSnapshot()
  })
})
