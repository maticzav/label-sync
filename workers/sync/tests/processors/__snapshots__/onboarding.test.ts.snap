// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onboarding onboards organization 1`] = `
Array [
  Object {
    "method": "check_installation_access",
    "params": Object {
      "owner": "test-org",
      "repos": Array [],
    },
  },
  Object {
    "method": "get_repo",
    "params": Object {
      "owner": "test-org",
      "repo": "test-org-labelsync",
    },
  },
  Object {
    "method": "bootstrap_config_repository",
    "params": Object {
      "repo": Object {
        "owner": "test-org",
        "repo": "test-org-labelsync",
      },
      "tree": Object {
        ".gitignore": "dist
node_modules

.DS_Store
*.log*",
        "README.md": "# LabelSync Yaml configuration.

Hey there! Welcome to LabelSync. We have scaffolded the configuration file for you. Check it out!

### Setting up LabelSync

1. Create a repository on Github and name it \`test-org-labelsync\`.
1. Commit your configuration (this repository) to Github.
1. Head over to [LabelSync Manager Github Application](https://github.com/apps/labelsync-manager) and make sure that you install it in all repositories that you have configured.
",
        "labelsync.yml": "# Colors
colors:
  area: &area \\"#FFD700\\"
  kind: &kind \\"#3B5BDB\\"
  status: &status \\"#F8F9FA\\"
  bug: &bug \\"#EE0000\\"
  priority: &priority \\"#F783AC\\"
  scope: &scope \\"#27CF79\\"
  team: &team'#FDF4E8'
  release: &release \\"#A5D8ff\\"
  process: &process \\"#EB9100\\"

# There's one YAML feature that you *absolutely* want to use in your config.
#
# It's references. You make a referencable value using &name, and reuse it
# by typing *name. If you are referencing an object (like \\"labels\\" below), use
# \\"<<: *common\\" to spread the values out (like \\"repos\\" below \\"labels\\").
#
# To learn more about the special YAML symbols, check out
# https://medium.com/@kinghuang/docker-compose-anchors-aliases-extensions-a1e4105d70bd.

# Labels
labels: &common
  \\"bug/0-needs-info\\":
    color: *bug
    description: More information is needed for reproduction.
    siblings: [\\"kind/bug\\"]
  \\"bug/1-repro-available\\":
    color: *bug
    description: A reproduction exists and needs to be confirmed.
    siblings: [\\"kind/bug\\"]
  \\"bug/2-confirmed\\":
    color: *bug
    description: We have confirmed that this is a bug.
    siblings: [\\"kind/bug\\"]
  \\"kind/bug\\":
    color: *bug
    description: A reported bug.
    alias: [\\"bug\\"]
  \\"kind/regression\\":
    color: *kind
    description: A reported bug in functionality that used to work before.
  \\"kind/feature\\":
    color: *kind
    description: A request for a new feature.
    alias: [\\"enhancement\\"]
  \\"kind/improvement\\":
    color: *kind
    description: An improvement to existing feature and code.
  \\"kind/docs\\":
    color: *kind
    description: A documentation change is required.
    alias: [\\"documentation\\"]
  \\"kind/discussion\\":
    color: *kind
    description: Discussion is required.
  \\"kind/question\\":
    color: *kind
    description: Developer asked a question. No code changes required.
    alias: [\\"question\\"]

# Repositories
repos:
  a:
    config:
      removeUnconfiguredLabels: false
    labels:
      {}
      # <<: *common
      # Add labels here.
  b:
    config:
      removeUnconfiguredLabels: false
    labels:
      {}
      # <<: *common
      # Add labels here.
  c:
    config:
      removeUnconfiguredLabels: false
    labels:
      {}
      # <<: *common
      # Add labels here.
",
      },
    },
  },
]
`;

exports[`onboarding onboards personal account 1`] = `
Array [
  Object {
    "method": "check_installation_access",
    "params": Object {
      "owner": "test-user",
      "repos": Array [],
    },
  },
  Object {
    "method": "get_repo",
    "params": Object {
      "owner": "test-user",
      "repo": "test-user-labelsync",
    },
  },
]
`;

exports[`onboarding skips onboarding when config repo exists 1`] = `
Array [
  Object {
    "method": "check_installation_access",
    "params": Object {
      "owner": "test-org",
      "repos": Array [],
    },
  },
  Object {
    "method": "get_repo",
    "params": Object {
      "owner": "test-org",
      "repo": "test-org-labelsync",
    },
  },
]
`;
